pgx/src/datum/numeric.rs

```diff
@@ -7,218 +7,225 @@ All rights reserved.
-                E: Error,
-            {
-                self.visit_string(v.to_owned())
-            }
-
-            #[inline]
-            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
-            where
-                E: Error,
-            {
-                // try to convert the provided String value into a Postgres Numeric Datum
-                // if it doesn't raise an conversion error, then we're good
-                PgTryBuilder::new(|| {
-                    // this might throw, but that's okay
-                    let datum = Numeric(v.clone()).into_datum().unwrap();
-
-                    unsafe {
-                        // and don't leak the 'inet' datum Postgres created
-                        pg_sys::pfree(datum.cast_mut_ptr());
-                    }
-
-                    // we have it as a valid String
-                    Ok(Numeric(v.clone()))
-                })
-                .catch_when(PgSqlErrorCode::ERRCODE_INVALID_TEXT_REPRESENTATION, |_| {
-                    Err(Error::custom(format!("invalid Numeric value: {}", v)))
-                })
-                .execute()
+impl Drop for AnyNumeric {
+    fn drop(&mut self) {
+        if self.need_pfree {
+            unsafe {
+                pg_sys::pfree(self.inner.cast());
             }
         }
-
-        deserializer.deserialize_any(NumericVisitor)
     }
 }
 
-impl From<i8> for Numeric {
-    fn from(val: i8) -> Self {
-        Numeric(val.to_string())
+impl Display for AnyNumeric {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        let numeric_out = unsafe {
+            direct_function_call::<&CStr>(pg_sys::numeric_out, vec![self.as_datum()]).unwrap()
+        };
+        let s = numeric_out.to_str().expect("numeric_out is not a valid UTF8 string");
+        fmt.write_str(s)
     }
 }
 
-impl From<i16> for Numeric {
-    fn from(val: i16) -> Self {
-        Numeric(val.to_string())
+impl Debug for AnyNumeric {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        write!(fmt, "AnyNumeric(inner: {self}, need_pfree:{})", self.need_pfree)
     }
 }
 
-impl From<i32> for Numeric {
-    fn from(val: i32) -> Self {
-        Numeric(val.to_string())
+impl<const P: u32, const S: u32> Display for Numeric<P, S> {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(fmt, "{}", self.0)
     }
 }
 
-impl From<i64> for Numeric {
-    fn from(val: i64) -> Self {
-        Numeric(val.to_string())
+#[inline(always)]
+pub(crate) const fn make_typmod(precision: u32, scale: u32) -> i32 {
+    let (precision, scale) = (precision as i32, scale as i32);
+    match (precision, scale) {
+        (0, 0) => -1,
+        (p, s) => ((p << 16) | s) + pg_sys::VARHDRSZ as i32,
     }
 }
 
-impl From<u8> for Numeric {
-    fn from(val: u8) -> Self {
-        Numeric(val.to_string())
-    }
+/// Which way does a [`AnyNumeric`] sign face?
+pub enum Sign {
+    Negative,
+    Positive,
+    Zero,
 }
```

Comment:
`AnyNumeric::sign` calls the `numeric_sign` function, [which can return NaN](https://github.com/postgres/postgres/blob/36e545cd051ffe58064c1f38faeb64f4feef4aec/src/backend/utils/adt/numeric.c#L1433-L1437). That variant should be in the enum.